File: server/storage.py
from typing import Dict, List, Optional
from datetime import datetime
import uuid
from schemas import User, UserCreate, Product, ChatMessage, ChatMessageCreate, SenderType, Size

class MemStorage:
    def __init__(self):
        self.users: Dict[int, User] = {}
        self.products: Dict[int, Product] = {}
        self.chat_messages: List[ChatMessage] = []
        self.current_user_id = 1
        
        # Initialize with some dummy data
        self._initialize_dummy_data()
    
    def _initialize_dummy_data(self):
        """Initialize storage with dummy data for testing"""
        # Add dummy users
        user_data = UserCreate(
            email="test@example.com",
            name="Test User"
        )
        user_id = self.current_user_id
        self.current_user_id += 1
        
        user = User(
            id=user_id,
            email=user_data.email,
            name=user_data.name
        )
        self.users[user_id] = user
        
        # Add dummy products
        self.products[1] = Product(
            id=1,
            title="Purina Pro Plan Adult Dog Food",
            brand="Purina",
            price=49.99,
            originalPrice=59.99,
            autoshipPrice=44.99,
            rating=4.5,
            reviewCount=1250,
            image="/images/dog-food-1.jpg",
            images=["/images/dog-food-1.jpg", "/images/dog-food-1-2.jpg"],
            deal=True,
            flavors=["Chicken & Rice", "Beef & Rice"],
            sizes=[
                Size(name="30 lb", price=49.99, pricePerLb="$1.67/lb"),
                Size(name="47 lb", price=69.99, pricePerLb="$1.49/lb")
            ],
            description="High-quality adult dog food with real chicken as the first ingredient.",
            inStock=True,
            category="Dog Food",
            keywords=["dog", "food", "adult", "chicken", "protein"]
        )
        
        # Add dummy chat messages
        self.chat_messages.append(ChatMessage(
            id=str(uuid.uuid4()),
            content="Hello! How can I help you find the perfect pet food today?",
            sender=SenderType.AI,
            timestamp=datetime.now()
        ))
    
    async def get_user(self, user_id: int) -> Optional[User]:
        """Get user by ID"""
        return self.users.get(user_id)
    
    async def get_user_by_email(self, email: str) -> Optional[User]:
        """Get user by email"""
        for user in self.users.values():
            if user.email == email:
                return user
        return None
    
    async def create_user(self, user_data: UserCreate) -> User:
        """Create a new user"""
        user_id = self.current_user_id
        self.current_user_id += 1
        
        user = User(
            id=user_id,
            email=user_data.email,
            name=user_data.name
        )
        
        self.users[user_id] = user
        return user
    
    async def get_products(self, category: Optional[str] = None, keywords: Optional[List[str]] = None) -> List[Product]:
        """Get products with optional filtering"""
        products = list(self.products.values())
        
        if category:
            products = [p for p in products if p.category.lower() == category.lower()]
        
        if keywords:
            # Simple keyword matching
            filtered_products = []
            for product in products:
                product_text = f"{product.title} {product.brand} {product.description}".lower()
                if any(keyword.lower() in product_text for keyword in keywords):
                    filtered_products.append(product)
            products = filtered_products
        
        return products
    
    async def get_product(self, product_id: int) -> Optional[Product]:
        """Get product by ID"""
        return self.products.get(product_id)
    
    async def get_chat_messages(self) -> List[ChatMessage]:
        """Get all chat messages"""
        return self.chat_messages
    
    async def add_chat_message(self, message_data: ChatMessageCreate) -> ChatMessage:
        """Add a new chat message"""
        message = ChatMessage(
            id=str(uuid.uuid4()),
            content=message_data.content,
            sender=message_data.sender,
            timestamp=datetime.now()
        )
        
        self.chat_messages.append(message)
        return message
    
    async def clear_chat_messages(self) -> None:
        """Clear all chat messages"""
        self.chat_messages.clear()

# Global storage instance
storage = MemStorage() 

File: server/storage.ts
import { User, Product, ChatMessage } from '../shared/schema';

export interface IStorage {
  // User operations
  getUserById(id: number): Promise<User | null>;
  getUserByEmail(email: string): Promise<User | null>;
  createUser(user: Omit<User, 'id'>): Promise<User>;
  
  // Product operations
  getAllProducts(): Promise<Product[]>;
  getProductById(id: number): Promise<Product | null>;
  searchProducts(query: string): Promise<Product[]>;
  getProductsByCategory(category: string): Promise<Product[]>;
  
  // Chat operations
  saveChatMessage(message: Omit<ChatMessage, 'id'>): Promise<ChatMessage>;
  getChatHistory(limit?: number): Promise<ChatMessage[]>;
}

export class MemStorage implements IStorage {
  private users: User[] = [];
  private products: Product[] = [];
  private chatMessages: ChatMessage[] = [];
  private nextUserId = 1;
  private nextProductId = 1;

  constructor() {
    // Initialize with mock data
    this.initializeMockData();
  }

  private initializeMockData() {
    // Add mock products
    this.products = [
      {
        id: 1,
        title: "Premium Dog Food - Adult Formula",
        brand: "Hill's Science Diet",
        price: 54.99,
        originalPrice: 64.99,
        autoshipPrice: 49.49,
        rating: 4.8,
        reviewCount: 1250,
        image: "/api/placeholder/300/300",
        images: ["/api/placeholder/300/300", "/api/placeholder/300/300"],
        deal: true,
        flavors: ["Chicken", "Beef", "Salmon"],
        sizes: [
          { name: "15 lb", price: 54.99, pricePerLb: "$3.67" },
          { name: "30 lb", price: 89.99, pricePerLb: "$3.00" }
        ],
        description: "Complete and balanced nutrition for adult dogs with high-quality protein and essential nutrients.",
        inStock: true,
        category: "Dog Food",
        keywords: ["premium", "nutrition", "adult", "dry food", "chicken", "protein"]
      },
      {
        id: 2,
        title: "Multi-Cat Clumping Litter",
        brand: "Fresh Step",
        price: 24.99,
        originalPrice: 29.99,
        autoshipPrice: 22.49,
        rating: 4.5,
        reviewCount: 890,
        image: "/api/placeholder/300/300",
        images: ["/api/placeholder/300/300"],
        deal: false,
        flavors: ["Unscented", "Fresh Scent"],
        sizes: [
          { name: "25 lb", price: 24.99, pricePerLb: "$1.00" }
        ],
        description: "Advanced odor control formula perfect for households with multiple cats.",
        inStock: true,
        category: "Cat Litter",
        keywords: ["odor control", "clumping", "multiple cats", "fresh", "unscented"]
      }
    ];

    this.nextProductId = this.products.length + 1;
  }

  // User operations
  async getUserById(id: number): Promise<User | null> {
    return this.users.find(user => user.id === id) || null;
  }

  async getUserByEmail(email: string): Promise<User | null> {
    return this.users.find(user => user.email === email) || null;
  }

  async createUser(userData: Omit<User, 'id'>): Promise<User> {
    const user: User = {
      ...userData,
      id: this.nextUserId++
    };
    this.users.push(user);
    return user;
  }

  // Product operations
  async getAllProducts(): Promise<Product[]> {
    return [...this.products];
  }

  async getProductById(id: number): Promise<Product | null> {
    return this.products.find(product => product.id === id) || null;
  }

  async searchProducts(query: string): Promise<Product[]> {
    const searchTerm = query.toLowerCase();
    return this.products.filter(product => 
      product.title.toLowerCase().includes(searchTerm) ||
      product.brand.toLowerCase().includes(searchTerm) ||
      product.category.toLowerCase().includes(searchTerm) ||
      product.keywords.some(keyword => keyword.toLowerCase().includes(searchTerm))
    );
  }

  async getProductsByCategory(category: string): Promise<Product[]> {
    return this.products.filter(product => 
      product.category.toLowerCase() === category.toLowerCase()
    );
  }

  // Chat operations
  async saveChatMessage(messageData: Omit<ChatMessage, 'id'>): Promise<ChatMessage> {
    const message: ChatMessage = {
      ...messageData,
      id: Date.now().toString()
    };
    this.chatMessages.push(message);
    return message;
  }

  async getChatHistory(limit: number = 50): Promise<ChatMessage[]> {
    return this.chatMessages
      .slice(-limit)
      .sort((a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime());
  }
}